test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.cc

```diff
@@ -239,64 +252,65 @@ DEFINE_PROTO_FUZZER(const test::extensions::filters::http::ext_proc::ExtProcGrpc
   FuzzedDataProvider ext_proc_provider(
       reinterpret_cast<const uint8_t*>(input.ext_proc_data().data()), input.ext_proc_data().size());
 
-  // Get IP and gRPC version from environment
-  ExtProcIntegrationFuzz fuzzer(TestEnvironment::getIpVersionsForTest()[0],
-                                TestEnvironment::getsGrpcVersionsForTest()[0]);
-  ExtProcFuzzHelper fuzz_helper(&ext_proc_provider);
-
-  // This starts an external processor in a separate thread. This allows for the
-  // external process to consume messages in a loop without blocking the fuzz
-  // target from receiving the response.
-  fuzzer.test_processor_.start(
-      fuzzer.ip_version_,
-      [&fuzz_helper](grpc::ServerReaderWriter<ProcessingResponse, ProcessingRequest>* stream) {
-        while (true) {
-          ProcessingRequest req;
-          if (!stream->Read(&req)) {
-            return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "expected message");
-          }
-
-          fuzz_helper.logRequest(&req);
-
-          // The following blocks generate random data for the 9 fields of the
-          // ProcessingResponse gRPC message
-
-          // 1 - 7. Randomize response
-          // If true, immediately close the connection with a random Grpc Status.
-          // Otherwise randomize the response
-          ProcessingResponse resp;
-          if (fuzz_helper.provider_->ConsumeBool()) {
-            ENVOY_LOG_MISC(trace, "Immediately Closing gRPC connection");
-            return fuzz_helper.randomGrpcStatusWithMessage();
-          } else {
-            ENVOY_LOG_MISC(trace, "Generating Random ProcessingResponse");
-            fuzz_helper.randomizeResponse(&resp, &req);
-          }
-
-          // 8. Randomize dynamic_metadata
-          // TODO(ikepolinsky): ext_proc does not support dynamic_metadata
-
-          // 9. Randomize mode_override
-          if (fuzz_helper.provider_->ConsumeBool()) {
-            ENVOY_LOG_MISC(trace, "Generating Random ProcessingMode Override");
-            ProcessingMode* msg = resp.mutable_mode_override();
-            fuzz_helper.randomizeOverrideResponse(msg);
+  // Initialize fuzzer once with IP and gRPC version from environment
+  if (fuzzer == nullptr) {
+    fuzzer = std::make_unique<ExtProcIntegrationFuzz>(
+        TestEnvironment::getIpVersionsForTest()[0], TestEnvironment::getsGrpcVersionsForTest()[0]);
+  }
+  // Initialize fuzz_helper every execution.
+  fuzz_helper.reset();
+  fuzz_helper = std::make_unique<ExtProcFuzzHelper>(&ext_proc_provider);
```

Comment:
Why is it static if it is recreated on every fuzzing iteration?

Reply:
Thanks for the comments!
fuzz_helper is used by the test client to generate random downstream request. It is also used by the test_server to generate the random mutated response.  Since the test_server is not created during each execution, the way for the test_server to get the current fuzz input is to use the process-wide global variable here.  The "static" keyword in front of the global variable is to keep it within this file.

Reply:
Also even in the persistent mode,  we are reset everything after 5000 fuzz executions to clean things up.  Since each Envoy creation + test server creation cost roughly ~500ms in a local VM.  Amortize it by 5000, it's 0.1ms/exec thus its effect is negligible in this mode.